cmd/kube-dns/SPEC.md

```diff
@@ -0,0 +1,99 @@
+    - The value of the `HostName` field stored in the map indexed by the endpoint’s IP address in the `endpoints.beta.kubernetes.io/hostnames-map` annotation.
+    - A unique, system-assigned identifier for the endpoint. The exact format and source of this identifier is not prescribed by this specification. However, it must be possible to use this to identify a specific endpoint in the context of a Service. This is used for SRV queries below in the event no explicit endpoint hostname is defined by the above methods.
+
+- _ready_
+
+ - An endpoint is considered _ready_ if its address is in the `addresses` field of the EndpointSubset object, or the corresponding service has the `service.alpha.kubernetes.io/tolerate-unready-endpoints` annotation set to `true`.
+
+The supported queries and expected responses follow.
+
+- **Type:** A, **Name:** `<service>.<ns>.svc.<zone>`
+
+  - If no Service named `<service>` exists in `<ns>`, the result will be `NXDOMAIN`.
+
+  - If a Service named `<service>` exists in `<ns>` with a ClusterIP, then the result will be a single A record with the specified name and the ClusterIP of that Service.
+
+  - If a Service named `<service>` exists in `<ns>` without a ClusterIP (i.e., a headless service)  and there are no _ready_ endpoints for that Service, then the result will be empty (i.e., an empty Answer section).
+
+  - If a Service named `<service>` exists in `<ns>` without a ClusterIP and there are _ready_ endpoints for the Service, the result will be an A record with the specified name and IP for each _ready_ endpoint.
+
+- **Type:** A, **Name:** `<hostname>.<service>.<ns>.svc.<zone>`
+
+  - If no Service named `<service>` exists in `<ns>`, the result will be `NXDOMAIN`.
+
+  - If a Service named `<service>` exists in `<ns>` and a _ready_ endpoint exists for that Service with _hostname_ of `<hostname>` the result will be a single A record with the IP address of that endpoint. If no such endpoint exists, the result will be `NXDOMAIN`.
+
+- **Type:** A, **Name:** `<a>-<b>-<c>-<d>.<ns>.pod.<zone>`
+
+  - If a Pod exists with IP address `<a>.<b>.<c>.<d>`, the result will be an A record with the specified name and the IP address `<a>.<b>.<c>.<d>`.
+
+  - If no such Pod exists, the result will be `NXDOMAIN`.
+
+- **Type:** SRV, **Name:** `_<port>._<proto>.<service>.<ns>.svc.<zone>`
+
+  - If no Service named `<service>` exists in `<ns>`, the result will be `NXDOMAIN`.
+
+  - If a Service named `<service>` exists in `<ns>` with a ClusterIP, and the service has a port named `<port>` using protocol `<proto>`, the result in the Answer section will be a single SRV record with the port number of the selected port and the name of the A record for the service (`<service>.<ns>.svc.<zone>`). If the service has exactly one port and it is unnamed, then `<port>` may be empty. The priority and weight returned are not prescribed by this specification. The Additional section may contain the named A record.
```

Comment:
I'm pretty sure if port is not specified, the entry is not generated in the current implementation